### PR TITLE
EZP-31462: Add handling unsupported password hash type by redirection to a password reset

### PR DESCRIPTION
**Pull request by kaff**
| Question      | Answer
| ------------- | ---
| Tickets       |  [EZP-31462](https://jira.ez.no/browse/EZP-31462)
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Tests pass?   | ?
| Doc needed?   | yes
| License       | [GPL-2.0](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/LICENSE)

In order to handle users with unsupported hash types redirection to "Password reset" was added

#### Checklist:
- [x] Implement tests
- [x] Coding standards (`$ composer fix-cs`)

### CODE REVIEW

File: src/bundle/Controller/PasswordResetController.php [hunk 1]
```diff
@@ -73,13 +73,11 @@ public function __construct(
...
 
     /**
-     * @param \Symfony\Component\HttpFoundation\Request $request
-     *
      * @return \EzSystems\EzPlatformUser\View\ForgotPassword\FormView|\EzSystems\EzPlatformUser\View\ForgotPassword\SuccessView|\Symfony\Component\HttpFoundation\RedirectResponse
      *
      * @throws \eZ\Publish\Core\Base\Exceptions\InvalidArgumentType
      */
-    public function userForgotPasswordAction(Request $request)
+    public function userForgotPasswordAction(Request $request, string $reason = null)
```
Comment (ViniTou) [inline src/bundle/Controller/PasswordResetController.php:80]:
```suggestion
    public function userForgotPasswordAction(Request $request, ?string $reason = null)
```

File: src/bundle/Type/UserForgotPasswordReason.php [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace EzSystems\EzPlatformUserBundle\Type;
+
+class UserForgotPasswordReason
+{
+    const MIGRATION = 'migration';
```
Comment (ViniTou) [inline src/bundle/Type/UserForgotPasswordReason.php:13]:
```suggestion
    public const MIGRATION = 'migration';
```

File: tests/Behat/Context/UserSetupContext.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+use Behat\Behat\Context\Context;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\ParameterType;
+
+class UserSetupContext implements Context
+{
+    private const UNSUPPORTED_USER_HASH = 5;
+
+    /** @var Connection */
```
Comment (ViniTou) [inline tests/Behat/Context/UserSetupContext.php:19]:
```suggestion
    /** @var \Doctrine\DBAL\Connection */
```

File: tests/Behat/Context/UserSetupContext.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    /**
+     * @Given a user :login has password in unsupported format
+     */
+    public function aUserHasPasswordInUnsupportedFormat($login)
```
Comment (ViniTou) [inline tests/Behat/Context/UserSetupContext.php:30]:
```suggestion
    public function aUserHasPasswordInUnsupportedFormat(string $login): void
```

File: .travis.yml [hunk 5]
```diff
@@ -28,13 +38,20 @@ before_install:
 - TEST_TIMEZONE=${TEST_TIMEZONES["`shuf -i 0-2 -n 1`"]}
 
 install:
-- travis_retry composer install --prefer-dist --no-interaction --no-suggest
+  - if [ "${CHECK_CS}" == "1" -o "${PHPUNIT_CONFIG}" != "" ]; then travis_retry composer install --prefer-dist --no-interaction --no-suggest ; fi
```
Comment (ViniTou) [inline .travis.yml:41]: I think beacuse you use `PHPUNIT_CONFIG` and `TEST_CONFIG` for running test (line 48) unit tests are failing on travis.

**Review by ViniTou — CHANGES_REQUESTED** [review]

File: .travis.yml [hunk 6]
```diff
@@ -28,13 +38,20 @@ before_install:
...
+  - if [ "${BEHAT_OPTS}" != "" ]; then ./.travis/prepare_ezplatform.sh ${INSTALL_EZ_INSTALL_TYPE}; fi
+
+before_script:
+  - if [ "${SETUP_BEHAT_OPTS}" != "" ] ; then cd "$HOME/build/ezplatform"; docker-compose exec --user www-data app sh -c "bin/ezbehat $SETUP_BEHAT_OPTS" ; fi
```
Comment (mnocon) [inline .travis.yml:45]: This part should be removed (we're not using SETUP_BEHAT_OPTS).

The `script` section should look like this:
```
script:
- if [ "$TEST_CONFIG" != "" ] ; then php -d date.timezone=$TEST_TIMEZONE -d memory_limit=-1
  vendor/bin/phpunit -c $TEST_CONFIG ; fi
- if [ "$CHECK_CS" == "1" ] ; then ./vendor/bin/php-cs-fixer fix -v --dry-run --show-progress=estimating; fi
- if [ "${BEHAT_OPTS}" != "" ]; then cd "$HOME/build/ezplatform"; docker-compose exec --user www-data app sh -c "bin/ezbehat ${BEHAT_OPTS}" ; fi
```

File: composer.json [hunk 7]
```diff
@@ -50,8 +52,10 @@
         "fix-cs": "@php ./vendor/bin/php-cs-fixer fix -v --show-progress=estimating"
     },
     "extra": {
+        "_ezplatform_branch_for_behat_tests": "master",
```
Comment (mnocon) [inline composer.json:55]:
```suggestion
        "_ezplatform_branch_for_behat_tests": "3.1",
```
We should use `master` for `master` branch, ezplatform-user 2.1 is used in Platform 3.1

File: .travis.yml [hunk 8]
```diff
@@ -16,6 +23,9 @@ matrix:
   - name: Code Style Check
     php: 7.3
     env: CHECK_CS=1
+  - name: "Workflow Integration tests"
```
Comment (mnocon) [inline .travis.yml:26]:
```suggestion
  - name: "User Integration tests"
```
Or something similar 😉

**Review by mnocon — CHANGES_REQUESTED** [review]
Small changes to Travis setup are needed - Behat part looks good! 😎

File: .travis.yml (hunk 5 above)
Comment (kaff) [reply .travis.yml:41]: fixed

File: behat_suites.yml [hunk 9]
```diff
@@ -0,0 +1,12 @@
+user:
+    suites:
+        browser:
+            paths:
+              - '%paths.base%/vendor/ezsystems/ezplatform-user/features/browser'
```
Comment (Nattfarinn) [inline behat_suites.yml:5]: Keep indent consistent within the file (4 spaces)

File: src/bundle/Resources/views/forgot_password/index.html.twig [hunk 10]
```diff
@@ -4,6 +4,9 @@
 
 {% block content %}
     <h2>{{ 'ezplatform.forgot_password.reset_your_password'|trans|desc('Reset your password') }}</h2>
+    {% if reason is defined and reason == constant('\\EzSystems\\EzPlatformUserBundle\\Type\\UserForgotPasswordReason::MIGRATION') %}
```
Comment (Nattfarinn) [inline src/bundle/Resources/views/forgot_password/index.html.twig:7]: Would be cleaner to inject constant value as parameter

**Review by Nattfarinn — APPROVED** [review]

File: behat_suites.yml (hunk 9 above)
Comment (kaff) [reply behat_suites.yml:5]: fixed

**Review by ViniTou — APPROVED** [review]

**Review by mnocon — COMMENTED** [review]
+1 for Travis/Behat

**Review by micszo — CHANGES_REQUESTED** [review]
As agreed in the call writing down found issues:

1. The exception `Argument 'hashType' is invalid: Password hash type '2' is not recognized` is now handled correctly when user attempts to log in (redirect happens) but it also occurs in the next step when user tries to change his password on the reset password page. The page just reloads without any message (which is a general issue, see e.g. https://jira.ez.no/browse/EZP-31854, https://jira.ez.no/browse/EZP-31304) but when user later logs in (with correct hash type) into admin panel the exception can be seen.
2. There are two paths in this case. User can try to log in on frontend or into admin. Both templates need to be updated with the new message about the password having expired. Namely: `/admin/user/reset-password/` and `/user/reset-password/`.

**Review by mikadamczyk — APPROVED** [review]

**Review by micszo — APPROVED** [review]
Retested successfully, the above issues do not occur anymore.
QA Approved on eZ Platform EE v3.1.2 with branches.

**Pull request merged**